### PR TITLE
Maine Main Income Tax

### LINKED ISSUE #2021: Maine Main Income Tax
Marginal tax on state taxable income (should be merged into master shortly!).

Legal code: https://www.mainelegislature.org/legis/statutes/36/title36sec5111.html
(this is fairly easy to read, use this!)

Examples:
NY main tax parameter folder (with one file for each filing status): https://github.com/PolicyEngine/policyengine-us/tree/a4052d137f7ac5753c7117b4fcaebdc300e60e39/policyengine_us/parameters/gov/states/ny/tax/income/main
NY main tax variable: https://github.com/PolicyEngine/policyengine-us/blob/a4052d137f7ac5753c7117b4fcaebdc300e60e39/policyengine_us/variables/gov/states/ny/tax/income/ny_main_income_tax.py

### PR DESCRIPTION
**Pull request by mzx-4real**
Fixes #2021

Co-authored-by: Leo <leogoldman@users.noreply.github.com>

### CODE REVIEW

File: policyengine_us/variables/gov/states/me/tax/income/me_main_income_tax.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from policyengine_us.model_api import *
+
+
+class me_main_income_tax(Variable):
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/me/tax/income/me_main_income_tax.py:4]: Could we call this `me_income_tax_before_credits`?

File: policyengine_us/parameters/gov/states/me/tax/income/main/head_of_household.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  threshold_unit: currency-USD
+  rate_unit: /1
+  name: me_income_tax_rate_hoh
+  label: ME State income tax rates for head-of-household filers
+  reference:
+    - title: Maine §5111. Imposition and rate of tax
+      href: https://www.mainelegislature.org/legis/statutes/36/title36sec5111.html
+brackets:
+  - threshold:
+      # Define time period and the first tax bracket lower bound
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/me/tax/income/main/head_of_household.yaml:13]: Could you remove the comments, or at least some?

File: policyengine_us/parameters/gov/states/me/tax/income/main/head_of_household.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+description: Progressive rate schedule for a head-of-household filer in ME. 
+metadata:
+  type: marginal_rate
+  threshold_unit: currency-USD
+  rate_unit: /1
+  name: me_income_tax_rate_hoh
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/me/tax/income/main/head_of_household.yaml:6]: We no longer need `name` attributes

File: policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+description: Progressive rate schedule for a separate filer in ME. 
+metadata:
+  type: marginal_rate
+  threshold_unit: currency-USD
+  rate_unit: /1
+  name: me_income_tax_rate_separate
+  label: ME State income tax rates for separate filers
+  reference:
+    - title: Maine §5111. Imposition and rate of tax
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml:9]: Could you please add a tax form reference as well? The law doesn't specify inflation adjustments but the form would confirm it.

File: policyengine_us/variables/gov/states/me/tax/income/me_income_tax_before_credits.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    unit = USD
+    definition_period = YEAR
+    defined_for = StateCode.ME
+
+    def formula(tax_unit, period, parameters):
+        taxable_income = tax_unit("me_taxable_income", period)
+        filing_status = tax_unit("filing_status", period)
+        status = filing_status.possible_values
+
+        rates = parameters(period).gov.states.me.tax.income.main
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/me/tax/income/me_income_tax_before_credits.py:17]: To adopt our standard
```suggestion
        p = parameters(period).gov.states.me.tax.income.main
```

File: policyengine_us/variables/gov/states/me/tax/income/me_income_tax_before_credits.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        single = rates.single
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/me/tax/income/me_income_tax_before_credits.py:18]: Let's do this inside the `select` to be more concise, e.g. `p.single.calc(taxable_income)`

File: policyengine_us/parameters/gov/states/me/tax/income/main/head_of_household.yaml (hunk 3 above)
Comment (MaxGhenis) [reply policyengine_us/parameters/gov/states/me/tax/income/main/head_of_household.yaml:6]: Other files too please

File: policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml (hunk 4 above)
Comment (mzx-4real) [reply policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml:9]: Hi, Max. I'm not so familiar with the tax form. So I might be forget to search for inflation adjustment. After looking up the tax form instruction and tax rate schedules, I do find out maine performs inflation adjustments. here is an example: [the inflation adjustment for 2023](https://www.maine.gov/revenue/sites/maine.gov.revenue/files/inline-files/ind_tax_rate_sched_2023.pdf). So basically I need to reflect those inflation adjustment by changing threshold and rate for each tax year in the yaml file?

Comment (leogoldman) [reply policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml:9]: @MaxGhenis are you saying there's something we should add for inflation adjustments? My understanding is that Maine updates the dollar thresholds to adjust for inflation, but the thresholds we have coded in here for 2022 are adjusted and there's no formula for calculating the next period's thresholds until they release the adjustment amount.

Below is the 2023 rates as well so we could also put those in though

Tax table: https://www.maine.gov/revenue/sites/maine.gov.revenue/files/inline-files/22_1040me_tax_tables.pdf
One-Pager for main tax rates, 2022: https://www.maine.gov/revenue/sites/maine.gov.revenue/files/inline-files/ind_tax_rate_sched_2022.pdf
One-Pager for main tax rates, 2023: https://www.maine.gov/revenue/sites/maine.gov.revenue/files/inline-files/ind_tax_rate_sched_2023.pdf
Tax Form (1040): https://www.maine.gov/revenue/sites/maine.gov.revenue/files/inline-files/22_1040me_dwnld_ff.pdf

Comment (MaxGhenis) [reply policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml:9]: The numbers here match the numbers starting in 2017 in the legal code. So if they're the same that means there's no inflation adjustment. I'd suggest adding the tax form as a reference to clarify/confirm this.

Yes if 2023 rules are available please include them.

File: policyengine_us/variables/gov/states/me/tax/income/me_income_tax_before_credits.py (hunk 6 above)
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/me/tax/income/me_income_tax_before_credits.py:18]: Please make sure to make the code changes before resolving conversations

File: policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml (hunk 4 above)
Comment (leogoldman) [reply policyengine_us/parameters/gov/states/me/tax/income/main/separate.yaml:9]: I see what you're saying!

@mzx-4real can you add the 2022 and 2023 threshold values from the forms linked above (and might as well leave the 2017 ones in!) and add a reference to the 2023 form too?

**Pull request merged**